databricks_cli/pipelines/api.py

```diff
@@ -38,28 +38,80 @@
 from databricks_cli.utils import InvalidConfigurationError
 
 BUFFER_SIZE = 1024 * 64
+base_pipelines_dir = 'dbfs:/pipelines/code/'
 
 
 class PipelinesApi(object):
     def __init__(self, api_client):
         self.client = DeltaPipelinesService(api_client)
         self.dbfs_client = DbfsApi(api_client)
 
+    def deploy(self, spec, headers=None):
+        """
+        Pipelines Deploy API
+        :param spec:
+        :param headers:
+        :return:
+        """
+        lib_objects = LibraryObject.convert_from_libraries(spec.get('libraries', []))
+        local_lib_objects, rest_lib_objects = \
+            self._partition_libraries_and_extract_local_paths(lib_objects)
+        remote_lib_objects = list(map(lambda llo:
+                                      LibraryObject(llo.lib_type, self._get_hashed_path(llo.path)),
+                                      local_lib_objects))
+        upload_files = self._get_files_to_upload(local_lib_objects, remote_lib_objects)
+
+        for llo, rlo in upload_files:
+            try:
+                self.dbfs_client.put_file(llo.path, rlo.path, False)
+            except Exception as e:
+                raise RuntimeError('Error \'{}\' while uploading {}'.format(e, llo.path))
+
+        rest_lib_objects.extend(remote_lib_objects)
+        spec['libraries'] = LibraryObject.convert_to_libraries(rest_lib_objects)
```

Comment:
Nit: We're not overloading rest_lib_objects to first be the non-upload files and later all the files? Can we write this functionally as LibraryObject.convert_to_libraries(rest_lib_objects + remote_lib_objects) ?

Reply:
This allows to save a copy, as extend modifies rest_lib_objects in place whereas + returns a new list. That being said, using + is more readable so I'm changing to that.